lib/resources/oidc.js

```diff
@@ -0,0 +1,184 @@
+// Copyright 2023 ODK Central Developers
+// See the NOTICE file at the top-level directory of this distribution and at
+// https://github.com/getodk/central-backend/blob/master/NOTICE.
+// This file is part of ODK Central. It is subject to the license terms in
+// the LICENSE file found in the top-level directory of this distribution and at
+// https://www.apache.org/licenses/LICENSE-2.0. No part of ODK Central,
+// including this file, may be copied, modified, propagated, or distributed
+// except according to the terms contained in the LICENSE file.
+
+// Allow declaring util functions at the end of the file:
+/* eslint-disable no-use-before-define */
+
+// OpenID Connect auth handling using Authorization Code Flow with PKCE.
+// TODO document _why_ auth-code-flow, and not e.g. implicit flow?
+
+const { generators } = require('openid-client');
+const config = require('config');
+
+const { html } = require('../util/html');
+const { redirect } = require('../util/http');
+const { createUserSession } = require('../util/sessions');
+const { // eslint-disable-line object-curly-newline
+  CODE_CHALLENGE_METHOD,
+  SCOPES,
+  getClient,
+  getRedirectUri,
+  isEnabled,
+} = require('../util/oidc'); // eslint-disable-line camelcase,object-curly-newline
+
+// TODO use req.protocol?
+const envDomain = config.get('default.env.domain');
+const HTTPS_ENABLED = envDomain.startsWith('https://');
+const ONE_HOUR = 60 * 60 * 1000;
+
+// Cannot use __Host- because cookie's Path is set
+// Use __Secure- in production.  But not in dev - even though firefox will
+// support __Secure with localhost, chrome will not.  Note that this behaviour
+// is similar but distinct from the Secure attribute, which seems to send
+// cookies to http://localhost on both Chrome and FireFox.
+// See:
+// * https://bugzilla.mozilla.org/show_bug.cgi?id=1648993
+// * https://bugs.chromium.org/p/chromium/issues/detail?id=1056543
+const CODE_VERIFIER_COOKIE = (HTTPS_ENABLED ? '__Secure-' : '') + 'ocv';
+const NEXT_COOKIE          = (HTTPS_ENABLED ? '__Secure-' : '') + 'next'; // eslint-disable-line no-multi-spaces
+const callbackCookieProps = {
+  httpOnly: true,
+  secure: HTTPS_ENABLED,
+  sameSite: 'Lax', // allow cookie to be sent on redirect from IdP
+  path: '/v1/oidc/callback',
+};
+
+module.exports = (service, endpoint) => {
+  if (!isEnabled()) return;
+
+  service.get('/oidc/login', endpoint.html(async ({ Sentry }, _, req, res) => {
+    try {
+      const client = await getClient();
+      const code_verifier = generators.codeVerifier(); // eslint-disable-line camelcase
+
+      const code_challenge = generators.codeChallenge(code_verifier); // eslint-disable-line camelcase
```

Comment:
can we make these camelCase?

Reply:
We can, but it complicates their use and dissociates the local variables from their corresponding API usage. Is that preferred?

Reply:
Usually, I think camelCase is much preferred over snake_case. However, I think @alxndrsn's explanation makes sense as to why this is an exception to the rule.